### PR TITLE
Feat/pn 7345

### PR DESCRIPTION
**Pull request by leleOFA**
## Short description
create logic and component to able cancel notification
## List of changes proposed in this pull request
- decomment part of notification detail with logic of cancel notificatio
- create new method and type for adapt cancel status of notification in timeline
- create new modal for understand cancel notification when payment is enabled 
- create test for all 

## How to test
- Login with grossini - palermo -> search for NRJP-NZRW-LDTL-202308-L-1 -> go to detail and check that notification has cancellation in progress
- Login with grossini - palermo -> search for HYTD-ERPH-WDUE-202308-H-1 -> go to detail and check that notification is cancelled
-  Login with grossini - palermo -> search for PELM-VYNK-XVGV-202308-R-1 -> go to detail and click on cancel button -> check that cancellation goes well
- Login with grossini - palermo -> search for XJTU-DJLJ-MEGE-202308-N-1 -> go to detail and click on cancel button -> check that cancellation returns 409
- Login with grossini - palermo -> go to detail (iun different from above) and click on cancel button -> check that cancellation returns 500

### CODE REVIEW

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 1]
```diff
@@ -275,25 +320,23 @@ const NotificationDetail = () => {
...
         ? t('detail.download-aar-expired', { ns: 'notifiche' })
         : t('detail.download-message-expired', { ns: 'notifiche' });
     },
     [isCancelled, hasDocumentsAvailable]
   );
 
-  // PN-1714
-  /*
   const openModal = () => {
+    console.log(' ciao ');
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:336]: remove console.log

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 2]
```diff
@@ -389,56 +406,32 @@ const NotificationDetail = () => {
...
-          {t('new-notification-button', { ns: 'notifiche' })}
-        </Button>
-      </DialogActions>
-    </Dialog>
-  );
-
   const direction = isMobile ? 'column-reverse' : 'row';
   const spacing = isMobile ? 3 : 0;
 
+  const apiError = hasNotificationSentApiError && (
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:412]: I don't think that it is needed. The complexity should be the same. Try to restore the previous version

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 3]
```diff
@@ -503,16 +496,19 @@ const NotificationDetail = () => {
...
               </Box>
             </Grid>
           </Grid>
         </Box>
       )}
-      <ModalAlert />
+      <ConfirmCancellationDialog
+        onClose={handleModalClose}
+        onConfirm={handleModalCloseAndProceed}
+        payment={true}
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:509]: do the real check

File: packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx [hunk 4]
```diff
@@ -67,6 +67,7 @@ describe('NotificationDetail Page (one recipient)', () => {
   let result: RenderResult;
   const mockDispatchFn = jest.fn();
   const mockActionFn = jest.fn();
+  const openModal = jest.fn();
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx:70]: open the modal simulating the click and after check that the modal is in the document

File: packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  const handleChange = () => {
+    setChecked(!checked);
+  };
+
+  useEffect(() => {
+    setChecked(false);
+  }, [onClose]);
+
+  return (
+    <Dialog open={showModal} data-testid="modalId">
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx:38]: Try to use the PnDialog. To understand how to use it, check around the app

File: packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  DialogTitle,
+  DialogContent,
+  DialogActions,
+  Button,
+  Checkbox,
+  FormControl,
+  FormControlLabel,
+  Typography,
+} from '@mui/material';
+import { t } from 'i18next';
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx:12]: not use t but the useTranslation hook

File: packages/pn-pa-webapp/src/redux/notification/actions.ts [hunk 7]
```diff
@@ -1,28 +1,41 @@
...
+}
 
 export const getSentNotification = createAsyncThunk<NotificationDetail, string>(
-  NOTIFICATION_ACTIONS.GET_SENT_NOTIFICATION, 
+  NOTIFICATION_ACTIONS.GET_SENT_NOTIFICATION,
   performThunkAction((params: string) => NotificationsApi.getSentNotification(params))
 );
 
-export const getSentNotificationLegalfact = createAsyncThunk<{url: string; retryAfter?: number}, {iun: string; legalFact: LegalFactId}>(
+export const cancelNotification = createAsyncThunk<CancelNotification, string>(
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/notification/actions.ts:28]: add the tests in reducers.test.tsx

File: packages/pn-pa-webapp/src/utils/AppError/GenericNotificationCancelledAppError.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import { AppError, ServerResponseError } from '@pagopa-pn/pn-commons';
+
+export class GenericNotificationCancelledAppError extends AppError {
+  private translateFunction: (path: string, ns: string) => string = (path: string) => path;
+
+  constructor(error: ServerResponseError, translateFunction: (path: string, ns: string) => string) {
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/utils/AppError/GenericNotificationCancelledAppError.ts:6]: Add the test

File: packages/pn-pa-webapp/src/utils/AppError/NotificationAlreadyCancelledAppError.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { AppError, ServerResponseError } from '@pagopa-pn/pn-commons';
+
+export class NotificationAlreadyCancelledAppError extends AppError {
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/utils/AppError/NotificationAlreadyCancelledAppError.ts:3]: Add the test

File: packages/pn-pa-webapp/src/utils/AppError/PAAppErrorFactory.ts [hunk 10]
```diff
@@ -25,6 +27,12 @@ export class PAAppErrorFactory extends AppErrorFactory {
         return new GenericInvalidParameterAppError(error, this.translateFunction);
       case ServerResponseErrorCode.PN_GENERIC_INVALIDPARAMETER_DUPLICATED:
         return new GenericInvalidParameterDuplicateAppError(error, this.translateFunction);
+      // capire se devo aggiungere qui un nuovo tipo di errore per l'annullamento
+      // e creare un nuovo ts dove creo il messaggio di errore
+      case ServerResponseErrorCode.PN_NOTIFICATION_ALREADY_CANCELLED:
+        return new NotificationAlreadyCancelledAppError(error, this.translateFunction);
+      case ServerResponseErrorCode.PN_GENERIC_CANCELLED_NOTIFICATION_ERROR:
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/utils/AppError/PAAppErrorFactory.ts:34]: Are you sure this will be the code of the error?

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

File: packages/pn-commons/src/components/NotificationDetail/NotificationDetailTimelineStep.tsx [hunk 11]
```diff
@@ -143,6 +143,9 @@ const NotificationDetailTimelineStep = ({
         label={notificationStatusInfos.label}
         color={position === 'first' ? notificationStatusInfos.color : 'default'}
         size={position === 'first' ? 'medium' : 'small'}
+        sx={{
+          opacity: notificationStatusInfos.tooltip.includes('corso') ? '0.5' : '1',
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-commons/src/components/NotificationDetail/NotificationDetailTimelineStep.tsx:147]: done

File: packages/pn-commons/src/types/NotificationDetail.ts [hunk 12]
```diff
@@ -339,6 +339,10 @@ export enum TimelineCategory {
   SEND_ANALOG_PROGRESS = 'SEND_ANALOG_PROGRESS',
   SEND_ANALOG_FEEDBACK = 'SEND_ANALOG_FEEDBACK',
   AAR_GENERATION = 'AAR_GENERATION',
+  NOTIFICATION_CANCELLATION_REQUEST = 'NOTIFICATION_CANCELLATION_REQUEST',
+  NOTIFICATION_CANCELLED = 'NOTIFICATION_CANCELLED',
+  /* CANCELLATION_IN_PROGRESS = 'CANCELLATION_IN_PROGRESS',
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-commons/src/types/NotificationDetail.ts:344]: done

File: packages/pn-commons/src/utils/TimelineUtils/TimelineStepFactory.ts [hunk 13]
```diff
@@ -40,7 +39,7 @@ export class TimelineStepFactory {
       case TimelineCategory.SEND_SIMPLE_REGISTERED_LETTER_PROGRESS:
         return new SendAnalogFlowStep();
       // -------------------------------------------
-        
+
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-commons/src/utils/TimelineUtils/TimelineStepFactory.ts:42]: done

File: packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx (hunk 6 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx:12]: done

File: packages/pn-pa-webapp/src/redux/notification/actions.ts (hunk 7 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/redux/notification/actions.ts:28]: done

File: packages/pn-pa-webapp/src/utils/AppError/GenericNotificationCancelledAppError.ts (hunk 8 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/utils/AppError/GenericNotificationCancelledAppError.ts:6]: done

File: packages/pn-pa-webapp/src/utils/AppError/NotificationAlreadyCancelledAppError.ts (hunk 9 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/utils/AppError/NotificationAlreadyCancelledAppError.ts:3]: done

File: packages/pn-pa-webapp/src/utils/AppError/PAAppErrorFactory.ts (hunk 10 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/utils/AppError/PAAppErrorFactory.ts:34]: I asked Simone Rondinella

File: packages/pn-commons/src/utils/__test__/notification.utility.test.ts [hunk 14]
```diff
@@ -422,10 +422,10 @@ describe('notification status texts', () => {
     testNotificationStatusInfosFnIncludingDescription(
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-commons/src/utils/__test__/notification.utility.test.ts]: done

File: packages/pn-pa-webapp/public/locales/it/notifiche.json [hunk 15]
```diff
@@ -438,5 +443,17 @@
         "message": "Il numero di protocollo inserito è già esistente"
       }
     }
+  },
+  "cancelled-notification": {
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/public/locales/it/notifiche.json:447]: done

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 16]
```diff
@@ -55,58 +42,56 @@ import {
...
-  );
-  const otherDocumentDownloadUrl = useAppSelector(
-    (state: RootState) => state.notificationState.otherDocumentDownloadUrl
-  );
-  const legalFactDownloadUrl = useAppSelector(
-    (state: RootState) => state.notificationState.legalFactDownloadUrl
-  );
-  const legalFactDownloadRetryAfter = useAppSelector(
-    (state: RootState) => state.notificationState.legalFactDownloadRetryAfter
+export const AlertNotificationCancel = (notification: {
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:56]: done

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 17]
```diff
@@ -55,58 +42,56 @@ import {
...
+  ) : (
+    <></>
   );
+};
+
+type Props = {
+  openModal: () => void;
+};
+
+export const RenderNotificationDetailTable = ({ openModal }: Props) => {
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:78]: done

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx (hunk 1 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx]: done

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx (hunk 2 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:415]: restore in previous version

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx (hunk 3 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx]: done

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:78]: i need to export the component for set openmodal fn on render in test

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 18]
```diff
@@ -275,25 +320,23 @@ const NotificationDetail = () => {
       if (isCancelled) {
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:324]: no, remove it

File: packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx [hunk 19]
```diff
@@ -55,58 +42,56 @@ import {
   getSentNotificationOtherDocument,
   getDowntimeLegalFactDocumentDetails,
   NOTIFICATION_ACTIONS,
+  cancelNotification,
 } from '../redux/notification/actions';
-import { setCancelledIun } from '../redux/newNotification/reducers';
+// import { setCancelledIun } from '../redux/newNotification/reducers';
```
[the remark this answers is not in the document]
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/NotificationDetail.page.tsx:48]: do it

File: packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx (hunk 5 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/components/Notifications/ConfirmCancellationDialog.tsx:40]: do

File: packages/pn-commons/src/utils/__test__/notification.utility.test.ts [hunk 20]
```diff
@@ -428,6 +428,17 @@ describe('notification status texts', () => {
       'status.canceled-description'
     );
   });
+
+  it('return notification status infos - CANCELLED - passing status only', () => {
```
Comment (AndreaCimini90) [inline packages/pn-commons/src/utils/__test__/notification.utility.test.ts:432]: Replace CANCELLED with CANCELLATION_IN_ PROGRESS

File: packages/pn-commons/src/utils/notification.utility.ts [hunk 21]
```diff
@@ -930,3 +945,28 @@ export function parseNotificationDetail(
   /* eslint-enable functional/no-let */
   return parsedNotification;
 }
+
+/**
+ * Insert cancelled status n timeline.
+ * @param  {NotificationDetail} notificationDetail
+ */
+const insertCancelledStatusInTimeline = (notificationDetail: NotificationDetail) => {
```
Comment (AndreaCimini90) [inline packages/pn-commons/src/utils/notification.utility.ts:953]: Put this definition before it's use

File: packages/pn-commons/src/utils/notification.utility.ts [hunk 22]
```diff
@@ -930,3 +945,28 @@ export function parseNotificationDetail(
...
+  const timelineCancelledElement = notificationDetail.timeline.find(
+    (el) => el.category === TimelineCategory.NOTIFICATION_CANCELLED
+  );
+
+  notificationDetail.timeline.forEach((element) => {
```
Comment (AndreaCimini90) [inline packages/pn-commons/src/utils/notification.utility.ts:958]: We can optimize this loop with this two steps:
- before loop check if timelineCancelledElement exists. if exists, DON'T do the loop
- instead a forEach, use find to get the NOTIFICATION_CANCELLATION_REQUEST and after push the new element

File: packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx [hunk 23]
```diff
@@ -67,6 +56,7 @@ describe('NotificationDetail Page (one recipient)', () => {
   let result: RenderResult;
   const mockDispatchFn = jest.fn();
   const mockActionFn = jest.fn();
+  // const openModal = jest.fn();
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx:59]: remove commented code

File: packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts [hunk 24]
```diff
@@ -148,4 +149,9 @@ describe('Notification detail redux state tests', () => {
     expect(state.legalFactDownloadRetryAfter).toEqual(0);
     expect(state.legalFactDownloadUrl).toEqual('');
   });
+
+  it('Should be able to cancel notification', async () => {
+    const action = await store.dispatch(cancelNotification('mocked-iun'));
+    expect(action.type).toBe('cancelNotification/rejected');
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts:155]: why the rejected check?

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

File: packages/pn-commons/src/utils/__test__/notification.utility.test.ts (hunk 20 above)
Comment (leleOFA) [reply packages/pn-commons/src/utils/__test__/notification.utility.test.ts:432]: do it

File: packages/pn-commons/src/utils/notification.utility.ts (hunk 21 above)
Comment (leleOFA) [reply packages/pn-commons/src/utils/notification.utility.ts:953]: do it

File: packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx (hunk 23 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/pages/__test__/NotificationDetail.page.test.tsx:59]: delete

File: packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts (hunk 24 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts:155]: I replace with fullfiled and skip the test

File: packages/pn-commons/src/utils/notification.utility.ts (hunk 22 above)
Comment (leleOFA) [reply packages/pn-commons/src/utils/notification.utility.ts]: do it

File: packages/pn-commons/src/utils/notification.utility.ts [hunk 25]
```diff
@@ -897,6 +911,33 @@ function timelineElementMustBeShown(t: INotificationDetailTimeline): boolean {
...
+/**
+ * Insert cancelled status n timeline.
+ * @param  {NotificationDetail} notificationDetail
+ */
+const insertCancelledStatusInTimeline = (notificationDetail: NotificationDetail) => {
+  const timelineCancelledElement = notificationDetail.timeline.find(
+    (el) => el.category === TimelineCategory.NOTIFICATION_CANCELLED
+  );
+  if (!timelineCancelledElement) {
+    notificationDetail.timeline.forEach((element) => {
```
Comment (AndreaCimini90) [inline packages/pn-commons/src/utils/notification.utility.ts:923]: the forEach is useless

File: packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts [hunk 26]
```diff
@@ -148,4 +149,9 @@ describe('Notification detail redux state tests', () => {
     expect(state.legalFactDownloadRetryAfter).toEqual(0);
     expect(state.legalFactDownloadUrl).toEqual('');
   });
+
+  it.skip('Should be able to cancel notification', async () => {
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts:153]: see the guide and mocks the api

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

File: packages/pn-commons/src/utils/notification.utility.ts (hunk 25 above)
Comment (leleOFA) [reply packages/pn-commons/src/utils/notification.utility.ts:923]: if i don'use for each, where i take element?

Comment (AndreaCimini90) [reply packages/pn-commons/src/utils/notification.utility.ts:923]: you are taking with the find

Comment (leleOFA) [reply packages/pn-commons/src/utils/notification.utility.ts]: do it

File: packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts (hunk 26 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/redux/notification/__test__/reducers.test.ts]: do it

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

**Comment by leleOFA** [comment]
iuns:  HYTD-ERPH-WDUE-202308-H-1 (Cancelled), NRJP-NZRW-LDTL-202308-L-1 (Cancellation in progress) 
user: grossini comune di palermo

File: packages/pn-pa-webapp/src/api/interceptors.ts [hunk 27]
```diff
@@ -1,16 +1,74 @@
-import { EnhancedStore } from "@reduxjs/toolkit";
-import { apiClient } from "./apiClients";
+import { EnhancedStore } from '@reduxjs/toolkit';
+// import { NotificationStatus } from '@pagopa-pn/pn-commons';
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/api/interceptors.ts:2]: remove commented code

File: packages/pn-pa-webapp/src/api/interceptors.ts [hunk 28]
```diff
@@ -1,16 +1,74 @@
...
+      if (token && config.headers) {
+        config.headers.Authorization = 'Bearer ' + token;
+      }
+      return config;
+    },
+    (error) => Promise.reject(error)
+  );
+
+  apiClient.interceptors.response.use(
+    (response) => {
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/api/interceptors.ts:20]: they are missed the cases:
- list of notifications
- all the cases of the cancelled api (success, already cancelled, generic error)

File: packages/pn-pa-webapp/src/api/notifications/Notifications.api.ts [hunk 29]
```diff
@@ -21,8 +26,10 @@ import {
   NOTIFICATION_PRELOAD_DOCUMENT,
   GET_USER_GROUPS,
   NOTIFICATION_DETAIL_OTHER_DOCUMENTS,
-} from './notifications.routes';
+  CANCEL_NOTIFICATION,
 
+  // CANCEL_NOTIFICATION,
```
Comment (AndreaCimini90) [inline packages/pn-pa-webapp/src/api/notifications/Notifications.api.ts:31]: remove commented code

**Review by AndreaCimini90 — CHANGES_REQUESTED** [review]

File: packages/pn-pa-webapp/src/api/interceptors.ts (hunk 27 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/api/interceptors.ts:2]: do it

File: packages/pn-pa-webapp/src/api/notifications/Notifications.api.ts (hunk 29 above)
Comment (leleOFA) [reply packages/pn-pa-webapp/src/api/notifications/Notifications.api.ts:31]: do it

**Comment by leleOFA** [comment]
i delete duplicated test, same test is done in notificationDetail.page

**Review by AndreaCimini90 — APPROVED** [review]

**Pull request merged**